File: implementation/swift/GameEventScriptTool/Sources/GameEventScriptTool/RunInventory.swift
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

import Foundation
import GameEventScriptRuntime

final class RunInventory {
    struct Entry {
        let id: Int
        let instance: GameEventScriptInstance
        let paths: [String]
    }
    let host: GameEventScriptHost
    let io: ToolIO
    private var entries: [Entry] = []
    private var native: [(String, GameEventScriptSubscription)] = []
    private(set) var nextID: Int
    var active: [Entry] { entries.filter { $0.instance.isAttached } }
    init(host: GameEventScriptHost, io: ToolIO, nextID: Int = 1) {
        self.host = host
        self.io = io
        self.nextID = nextID
    }
    @discardableResult func load(_ program: GameEventScriptProgram, paths: [String], id: Int? = nil) throws
        -> GameEventScriptInstance
    {
        let instance = try host.load(program)
        let assignedID = id ?? nextID
        entries.append(Entry(id: assignedID, instance: instance, paths: paths))
        nextID = max(nextID, assignedID + 1)
        return instance
    }
    func unload(_ selector: String) throws {
        let entry = try select(selector, command: ":unload")
        entry.instance.detach()
        entries.removeAll { $0.id == entry.id }
    }
    @discardableResult func unloadAll() -> Int {
        let count = active.count
        for entry in entries { entry.instance.detach() }
        entries.removeAll()
        return count
    }
    func subscribe(_ name: String, handler: any GameEventScriptNativeMessageHandler) throws {
        native.append((name, try host.subscribeMessageName(name, handler: handler)))
    }
    func handlers(_ program: GameEventScriptProgram) -> [GameEventScriptBinding] {
        program.bindings.filter { isHandler($0) && program.stringConstants[Int($0.name)] != "initialization" }
    }
    func programs() {
        io.line(toError: true)
        io.line("Loaded programs (\(active.count)):", toError: true)
        for entry in active {
            let p = entry.instance.program
            io.line(
                "  @\(entry.id)  \(p.moduleName)  version=\(p.programVersion)  handlers=\(handlers(p).count)",
                toError: true)
            io.line("      files: " + entry.paths.map(quoted).joined(separator: ", "), toError: true)
        }
        if active.isEmpty { io.line("  No programs loaded. Use :load <file>.", toError: true) }
        io.line(toError: true)
    }
    func registeredHandlers() {
        var lines = native.filter { $0.1.isSubscribed }.map { "  native  \($0.0)(...) [name-only]" }
        for entry in active {
            let p = entry.instance.program
            func tags(_ prefix: String, _ values: [UInt16]) -> String {
                values.isEmpty ? "" : prefix + values.map { "#" + p.stringConstants[Int($0)] }.joined(separator: ", ")
            }
            for binding in handlers(p) {
                let name = p.stringConstants[Int(binding.name)]
                let signature =
                    binding.kind == .messageNameHandler
                    ? name + "(...) [name-only]"
                    : name + "(" + binding.argumentNames.map { p.stringConstants[Int($0)] }.joined(separator: ",")
                        + ") [signature]"
                lines.append(
                    "  @\(entry.id)  \(p.moduleName)  \(signature)" + tags(" matching ", binding.requiredTags)
                        + tags(" without ", binding.excludedTags))
            }
        }
        io.line(toError: true)
        io.line("Registered handlers (\(lines.count)):", toError: true)
        for line in lines { io.line(line, toError: true) }
        io.line(toError: true)
    }
    func select(_ selector: String, command: String) throws -> Entry {
        guard !selector.isEmpty, !selector.contains(where: \.isWhitespace) else {
            throw ToolError.usage("Specify one module name or @program-id after \(command). Use :list.")
        }
        let matches: [Entry]
        if selector.hasPrefix("@") {
            let value = String(selector.dropFirst())
            guard RunOptions.integerSpelling(value, signed: false), let id = Int(value), id > 0 else {
                throw ToolError.usage("Use @ followed by a positive integer, for example \(command) @1.")
            }
            matches = active.filter { $0.id == id }
        } else {
            matches = active.filter { GesText.scalarEqual($0.instance.program.moduleName, selector) }
        }
        guard !matches.isEmpty else { throw ToolError.usage("No loaded program matches '\(selector)'. Use :list.") }
        guard matches.count == 1 else {
            throw ToolError.usage(
                "Module '\(selector)' is loaded more than once. Use \(command) with one of: "
                    + matches.map { "@\($0.id)" }.joined(separator: ", ") + ".")
        }
        return matches[0]
    }
    func dump(_ selector: String, color: Bool) throws {
        let p = try select(selector, command: ":dump").instance.program
        var text = GameEventScriptProgramDumper.dump(p)
        if io.errorTerminal { text = TextDisplay.expandTabs(text) }
        io.line(toError: true)
        io.write(color ? Highlighting().renderAssembly(text) : text, toError: true)
        io.line(toError: true)
    }
    func source(_ selector: String, color: Bool) throws {
        let entry = try select(selector, command: ":source")
        io.line(toError: true)
        guard let sources = entry.instance.program.sourceArchive, !sources.isEmpty else {
            io.line(
                "No embedded sources available for @\(entry.id) (\(entry.instance.program.moduleName)). Compile with source archive/debug information to include them.",
                toError: true)
            io.line(toError: true)
            return
        }
        for source in sources {
            let header = "// Source: " + quoted(source.sourceName)
            io.line(color ? Highlighting.paint(header, 90) : header, toError: true)
            var text = source.text
            if io.errorTerminal { text = TextDisplay.expandTabs(text) }
            io.write(color ? Highlighting().render(text) : text, toError: true)
            if text.utf8.last != 10 && text.utf8.last != 13 { io.line(toError: true) }
            io.line(toError: true)
        }
    }
}

File: implementation/csharp/GameEventScript.Tool/src/RunInventory.cs
// Copyright 2026 Stephan Schlöpke
// SPDX-License-Identifier: Apache-2.0

using System.Globalization;
using System.Text.Json;
using GameEventScript.Api;
using static GameEventScript.Api.GameEventScriptBindingSegment;

namespace GameEventScript.Tool;

internal sealed class RunInventory(GameEventScriptHost host, int nextId = 1)
{
    private readonly List<LoadedProgram> _programs = [];
    private readonly List<NativeHandler> _nativeHandlers = [];

    internal int NextId { get; private set; } = nextId;

    internal GameEventScriptInstance Load(GameEventScriptProgram program, IReadOnlyList<string> paths, int? id = null)
    {
        var instance = host.Load(program);
        var assignedId = id ?? NextId;
        _programs.Add(new LoadedProgram(assignedId, instance, [.. paths]));
        NextId = Math.Max(NextId, assignedId + 1);
        return instance;
    }

    internal void Unload(string selector)
    {
        var entry = SelectProgram(selector, ":unload");
        entry.Instance.Detach();
        _programs.Remove(entry);
    }

    internal int UnloadAll()
    {
        var count = ActivePrograms().Length;
        foreach (var entry in _programs) entry.Instance.Detach();
        _programs.Clear();
        return count;
    }

    internal GameEventScriptSubscription SubscribeMessageName(string name, IGameEventScriptNativeMessageHandler handler)
    {
        var subscription = host.SubscribeMessageName(name, handler);
        _nativeHandlers.Add(new NativeHandler(name, subscription));
        return subscription;
    }

    internal void WritePrograms()
    {
        var programs = ActivePrograms();
        Console.Error.WriteLine();
        Console.Error.WriteLine($"Loaded programs ({programs.Length}):");
        foreach (var entry in programs)
        {
            Console.Error.WriteLine($"  @{entry.Id}  {entry.Instance.Program.ModuleName}  version={entry.Instance.Program.ProgramVersion}  handlers={Handlers(entry).Count()}");
            Console.Error.WriteLine("      files: " + string.Join(", ", entry.Paths.Select(path => JsonSerializer.Serialize(path))));
        }
        if (programs.Length == 0) Console.Error.WriteLine("  No programs loaded. Use :load <file>.");
        Console.Error.WriteLine();
    }

    internal void WriteHandlers()
    {
        var lines = new List<string>();
        foreach (var handler in _nativeHandlers)
        {
            if (handler.Subscription.IsSubscribed) lines.Add($"  native  {handler.Name}(...) [name-only]");
        }
        foreach (var entry in ActivePrograms())
        {
            var strings = entry.Instance.Program.StringConstants;
            foreach (var binding in Handlers(entry))
            {
                var name = strings.Resolve(binding.Name);
                var signature = binding.Kind == GameEventScriptBinaryBindKind.MessageNameHandler
                    ? name + "(...) [name-only]"
                    : GameEventScriptMessageSignature.CreateSignatureId(name, binding.ArgumentNames.Select(index => strings.Resolve(index))) + " [signature]";
                var tags = Tags(" matching ", binding.RequiredTags, strings) + Tags(" without ", binding.ExcludedTags, strings);
                lines.Add($"  @{entry.Id}  {entry.Instance.Program.ModuleName}  {signature}{tags}");
            }
        }
        Console.Error.WriteLine();
        Console.Error.WriteLine($"Registered handlers ({lines.Count}):");
        foreach (var line in lines) Console.Error.WriteLine(line);
        Console.Error.WriteLine();
    }

    internal void WriteDump(string selector, bool color)
    {
        var text = SelectProgram(selector, ":dump").Instance.Program.Dump();
        if (!Console.IsErrorRedirected) text = ToolTextDisplay.ExpandTabs(text);
        Console.Error.WriteLine();
        Console.Error.Write(color ? RunHighlighting.RenderAssembly(text) : text);
        Console.Error.WriteLine();
    }

    internal void WriteSource(string selector, bool color)
    {
        var entry = SelectProgram(selector, ":source");
        Console.Error.WriteLine();
        if (entry.Instance.Program.SourceArchive is not { Sources.Count: > 0 } archive)
        {
            Console.Error.WriteLine($"No embedded sources available for @{entry.Id} ({entry.Instance.Program.ModuleName}). Compile with source archive/debug information to include them.");
            Console.Error.WriteLine();
            return;
        }
        foreach (var source in archive.Sources)
        {
            var header = "// Source: " + JsonSerializer.Serialize(source.SourceName);
            Console.Error.WriteLine(color ? RunHighlighting.Paint(header, 90) : header);
            var text = source.ResolveText();
            if (!Console.IsErrorRedirected) text = ToolTextDisplay.ExpandTabs(text);
            Console.Error.Write(color ? RunHighlighting.Render(text) : text);
            if (!text.EndsWith('\n') && !text.EndsWith('\r')) Console.Error.WriteLine();
            Console.Error.WriteLine();
        }
    }

    private LoadedProgram SelectProgram(string selector, string command)
    {
        if (selector.Length == 0 || selector.Any(char.IsWhiteSpace)) throw new ArgumentException($"Specify one module name or @program-id after {command}. Use :list to see loaded programs.");
        var programs = ActivePrograms();
        LoadedProgram[] matches;
        if (selector.StartsWith('@'))
        {
            if (!int.TryParse(selector.AsSpan(1), NumberStyles.None, CultureInfo.InvariantCulture, out var id) || id <= 0)
                throw new ArgumentException($"A program selector uses @ followed by a positive integer, for example {command} @1.");
            matches = programs.Where(entry => entry.Id == id).ToArray();
        }
        else matches = programs.Where(entry => string.Equals(entry.Instance.Program.ModuleName, selector, StringComparison.Ordinal)).ToArray();
        if (matches.Length == 0) throw new ArgumentException($"No loaded program matches '{selector}'. Use :list.");
        if (matches.Length > 1)
        {
            var choices = string.Join(", ", matches.Select(entry => $"@{entry.Id}"));
            throw new ArgumentException($"Module '{selector}' is loaded more than once. Use {command} with one of: {choices}.");
        }
        return matches[0];
    }

    internal LoadedProgram[] ActivePrograms() => _programs.Where(entry => entry.Instance.IsAttached).ToArray();

    private static IEnumerable<GameEventScriptBinaryBindEntry> Handlers(LoadedProgram entry)
        => entry.Instance.Program.Bindings.Entries.Where(binding => binding.Kind is GameEventScriptBinaryBindKind.MessageHandler or GameEventScriptBinaryBindKind.MessageNameHandler
            && entry.Instance.Program.StringConstants.Resolve(binding.Name) != "initialization");

    private static string Tags(string prefix, GameEventScriptReadOnlyArray<ushort> tags, GameEventScriptStringConstantSegment strings)
        => tags.Count == 0 ? string.Empty : prefix + string.Join(", ", tags.Select(index => "#" + strings.Resolve(index)));

    internal sealed record LoadedProgram(int Id, GameEventScriptInstance Instance, string[] Paths);
    private sealed record NativeHandler(string Name, GameEventScriptSubscription Subscription);
}
